src/main/scala/analysis/NonReturningFunctions.scala

```diff
@@ -0,0 +1,139 @@
+package analysis
+
+import bap.{BAPJump, BAPSubroutine}
+import ir.{Block, DirectCall, GoTo, IndirectCall, Jump, Procedure, Statement}
+import specification.ExternalFunction
+
+import scala.collection.mutable.{Map, Queue}
+import collection.parallel.CollectionConverters.seqIsParallelizable
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.parallel.CollectionConverters.*
+
+class NonReturningFunctions {
+
+  def transform(procedures: ArrayBuffer[Procedure]): Unit = {
+    val blocksToRemove: Queue[String] = Queue()
+    val mapJumpsToBlocks: Map[String, ArrayBuffer[(Jump, Block)]] = Map()
+    val mapBlocksToProcedure: Map[String, (Procedure, Integer)] = Map()
+
+
+    // Check if the goto is part of a endless loop by checking to see if the index its jumping to is earlier on in the method,
+    // and that any jumps between the index its jumping to, and the index of the instruction, don't jump outside of that range.
+    // This method checks for both continue (which jumps back to the while loop, and break (which jumps outside of loops).
+    def isEndlessLoop(proc: Procedure, goTo: GoTo, index: Integer): Boolean = {
+
+      if (goTo.condition.isEmpty && mapBlocksToProcedure.contains(goTo.target.label) && mapBlocksToProcedure(goTo.target.label)._2 < index) {
+
+        val (_, idx) = mapBlocksToProcedure(goTo.target.label)
+
+        for (loopIndex <- idx.toInt to index.toInt) {
+          val blockAtLoopIndex = proc.blocks(loopIndex)
+          for (loopJump <- blockAtLoopIndex.jumps) {
+            loopJump match {
+              case loopGoTo: GoTo =>
+                if (!mapBlocksToProcedure.contains(loopGoTo.target.label) || mapBlocksToProcedure(loopGoTo.target.label)._2 >= index)
+                  return false

```

Comment:
mapBlocksToProcedure associates blocks with an index based on the order they are iterated over. Using this is not a safe way to identify the control flow between blocks. The only guarantee we have about the order of blocks is that the first block in a procedure is the entry point.

Reply:
Actually according to code and documentation, (see git repository), you can rely on the order of the blocks being the order in which they apear in the procedure. Meaning that the code may be in different locations, but BAP will reorder these blocks so they are in the same order as the orginal program.

The IR to boogie, also relies on this approache as well.

Reply:
Where does anything say that you can rely on the order of the blocks in this way? If it says that, it's incorrect.

There is a difference between the order of the blocks and the control flow of the program. If we have a procedure with blocks A, B, C  in that order, the only guarantee is that block A is the first block in the procedure. Control flow between the blocks is determined entirely by jumps. It is not reliant on the order of blocks at all beyond the first block being the entry point.

Reply:
> Where does anything say that you can rely on the order of the blocks in this way? If it says that, it's incorrect.
> 
> There is a difference between the order of the blocks and the control flow of the program. If we have a procedure with blocks A, B, C in that order, the only guarantee is that block A is the first block in the procedure. Control flow between the blocks is determined entirely by GoTos. It is not reliant on the order of blocks at all beyond the first block being the entry point.

See https://github.com/BinaryAnalysisPlatform/bap/blob/master/plugins/print/print_main.ml line 207 and
https://github.com/BinaryAnalysisPlatform/bap/blob/master/lib/bap_types/bap_ir.ml see lines 1888, 1861 and 1853 

BAP itself relies on this fact during other areas of its source code as well (available on request).

> 
> IRToBoogie doesn't rely on the blocks being in any particular order either.

Lines 285, and 379 implicitly rely on the same order the instructions are found in the BAPToIR translation.

Also there isn't really any other way to remove code that can't be jumped to once a endless while loop is found. The fact that I have found an endless loop means there isn't any control flow to the next block, so even using the CFG will have the same issues, the only difference is the CFG will make it much more difficult (both in time and memory costs) to find endless loops.

Hope that helps.

Reply:
I don't see evidence in those examples of the order of the blocks implying anything about the control flow. BAP may order the blocks in a certain way, but we are not relying on such intricacies of BAP where possible, as there are intended to be inputs other than BAP at some point. 

IRToBoogie does not rely on the blocks being in any particular order inside a procedure there, beyond the initial block being first, because that is all that matters to Boogie. Control flow between blocks is entirely dependent on GoTos.

I don't know what you mean by 'using the CFG will have the same issues'? I'm not suggesting that CFG should be used to identify infinite loops. Removing unreachable code should happen in a separate method to identifying non-returning procedures and marking calls as non-returning, because we don't necessarily want to do both at the same time.

Reply:
> I don't see evidence in those examples of the order of the blocks implying anything about the control flow. BAP may order the blocks in a certain way, but we are not relying on such intricacies of BAP where possible, as there are intended to be inputs other than BAP at some point.
> 

I can assure you those are the lines that indiccate the order of the control flow (rest of the proof comes from many other files, and don't really have time to prove it entirely). Also its own implementation relies on this fact as well.

> 
> I don't know what you mean by 'using the CFG will have the same issues'? I'm not suggesting that CFG should be used to identify infinite loops. Removing unreachable code should happen in a separate method to identifying non-returning procedures and marking calls as non-returning, because we don't necessarily want to do both at the same time.

We do actually want it to be in the same process. The effect of removing code could actually render a function non-returning.

and example is basic_function_calling_nonreturning_function.c. The function_calling_non_returning_function has a return statement, which is only considered non-returning because the code return statement is dead code. Therefore each time code is removed, an evaluation is needed to determine if the function is still non-returning.

The issue with while loops is there is by definition, no way to identify where the next block is. This is because if there was a jump to the next block then there would be a control path that jumps, and thus making the function a returning function. Therefore this code needs to know the order of the blocks. Also this needs to be done at this level to remove code that could return but is not reachable.

The other way you could do this is to use the CFG, however I'm not sure this would even work either, as I can't simply just traverse the CFG. Also from what I understand while loops is something the CFG has problems with currently.

If your concerned about it, I can remove the detection of endless loops entirely, but this would mean that some functions that are non-returning will be marked as non-returning.

Reply:
It doesn't matter if BAP is implicitly guaranteeing anything about the control flow in relation to the order of blocks. We are not relying on such implicit guarantees from BAP in the IR, because that means we have to maintain said guarantee whenever we modify the IR, and there is also the intention to add support for lifters other than BAP.

In order to remove unreachable blocks, you could simply iterate over the blocks in the procedure and remove the ones that do not have jumps to them (except for the initial block). This could easily be separated from identifying non-returning procedures and called later in the program. 

Dealing with infinite loops is not very important from my perspective, since we do not currently fully support loops in any form for a variety of reasons.

> The issue with while loops is there is by definition, no way to identify where the next block is. This is because if there was a jump to the next block then there would be a control path that jumps, and thus making the function a returning function. Therefore this code needs to know the order of the blocks.

The order blocks are contained within the procedure is irrelevant to reachability except for the first block, which is always reachable. It is possible to identify unreachable blocks without this idea of a 'next block'.

> We do actually want it to be in the same process. The effect of removing code could actually render a function non-returning.
> 
> and example is basic_function_calling_nonreturning_function.c. The function_calling_non_returning_function has a return statement, which is only considered non-returning because the code return statement is dead code. Therefore each time code is removed, an evaluation is needed to determine if the function is still non-returning.

It doesn't seem like this happens though, it seems like it stops trying to detect non-returning calls as soon as blocks have been deleted? There is also a difference between the block being unreachable and being removed. It would be possible to determine if the returns are now unreachable after setting returnTarget to None without removing the unreachable blocks completely.

Marking calls as non-returning is fine to happen when you identify non-returning procedures, but I want to keep removing the unreachable blocks separate and have it happen at the very end of the program, in case later analysis finds that the blocks are in fact reachable in some other way, for instance. I can anticipate needing to consider the order these things happen in and do not want them to be tied so directly together.

Reply:
> It doesn't matter if BAP is implicitly guaranteeing anything about the control flow in relation to the order of blocks. We are not relying on such implicit guarantees from BAP in the IR, because that means we have to maintain said guarantee whenever we modify the IR, and there is also the intention to add support for lifters other than BAP.
> 
> In order to remove unreachable blocks, you could simply iterate over the blocks in the procedure and remove the ones that do not have jumps to them (except for the initial block). This could easily be separated from identifying non-returning procedures and called later in the program.
> 
> Dealing with infinite loops is not very important from my perspective, since we do not currently fully support loops in any form for a variety of reasons.
> 

I'll remove the code that looks at infinite loops, athough there are issues here that need to be addressed, if we are truely going to identify all non-returning functions. I'll leave it to someone else to implement.

> 
> The order blocks are contained within the procedure is irrelevant to reachability except for the first block, which is always reachable. It is possible to identify unreachable blocks without this idea of a 'next block'.

Lets do that in another ticket.


> 
> > We do actually want it to be in the same process. The effect of removing code could actually render a function non-returning.
> > and example is basic_function_calling_nonreturning_function.c. The function_calling_non_returning_function has a return statement, which is only considered non-returning because the code return statement is dead code. Therefore each time code is removed, an evaluation is needed to determine if the function is still non-returning.
> 
> It doesn't seem like this happens though, it seems like it stops trying to detect non-returning calls as soon as blocks have been deleted? There is also a difference between the block being unreachable and being removed. It would be possible to determine if the returns are now unreachable after setting returnTarget to None without removing the unreachable blocks completely.


line 88 in NonReturningFunctions.scala loops until there isn't any deleted blocks.

Lets talk about whether we should deleted the blocks unreachable on Tuesday.

> 
> Marking calls as non-returning is fine to happen when you identify non-returning procedures, but I want to keep removing the unreachable blocks separate and have it happen at the very end of the program, in case later analysis finds that the blocks are in fact reachable in some other way, for instance. I can anticipate needing to consider the order these things happen in and do not want them to be tied so directly together.

The way I handle reachability is not to always delete the block but to determine whether the code can infact be jumped too by other blocks. If it can, then it isn't deleted (i.e that is where it stops).

Seperated the deleteing from the marking of deleting.

Any further changes should be in another ticket.